lib/xnvme_be_macos_driverkit_async.c

```diff
@@ -0,0 +1,202 @@
+	mach_vm_size_t size = 0;
+	mach_vm_address_t addr;
+	kern_return_t res;
+	int qpid;
+
+	qpid = _xnvme_be_driverkit_create_ioqpair(state, q->base.capacity, 0);
+	if (qpid < 0) {
+		XNVME_DEBUG("FAILED: nvme_create_ioqpair(); err: %d", qpid);
+		return qpid;
+	}
+	q->id = qpid;
+
+	res = IOConnectMapMemory64(state->device_connection, 1 << 16 | qpid, mach_task_self(),
+				   &addr, &size, kIOMapAnywhere);
+	XNVME_DEBUG("IOConnectMapMemory64 returned: %x, %s", res, mach_error_string(res));
+	assert(res == kIOReturnSuccess);
+	q->sq_queue = (RingQueue *)addr;
+	assert(q->sq_queue->buffer_size == size);
+
+	size = 0;
+	res = IOConnectMapMemory64(state->device_connection, qpid, mach_task_self(), &addr, &size,
+				   kIOMapAnywhere);
+	XNVME_DEBUG("IOConnectMapMemory64 returned: %x, %s", res, mach_error_string(res));
+	assert(res == kIOReturnSuccess);
+	q->cq_queue = (RingQueue *)addr;
+	assert(q->cq_queue->buffer_size == size);
+
+	return 0;
+}
+
+int
+xnvme_be_macos_driverkit_queue_term(struct xnvme_queue *_queue)
+{
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)_queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)q->base.dev->be.state;
+
+	_xnvme_be_driverkit_delete_ioqpair(state, q->id);
+
+	q->cq_queue = NULL;
+	q->sq_queue = NULL;
+	return 0;
+}
+
+int
+xnvme_be_macos_driverkit_queue_poke(struct xnvme_queue *queue, uint32_t XNVME_UNUSED(max))
+{
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)queue->base.dev->be.state;
+	uint64_t qid[2] = {q->id, q->id}; // Submission queue, completion queue
+	int reaped = 0;
+	int queue_ret;
+	NvmeSubmitCmd nvme_cmd;
+	kern_return_t ret;
+	struct xnvme_cmd_ctx *ctx;
+
+	ret = IOConnectCallScalarMethod(state->device_connection, NVME_POKE, qid, 2, NULL, 0);
+	if (ret != kIOReturnSuccess) {
+		return -1;
+	}
+	while ((queue_ret = queue_dequeue(q->cq_queue, &nvme_cmd)) == 0) {
+		ctx = (struct xnvme_cmd_ctx *)nvme_cmd.backend_opaque;
+		memcpy(&ctx->cpl, nvme_cmd.cpl, sizeof(struct xnvme_spec_cpl));
+		ctx->async.cb(ctx, ctx->async.cb_arg);
+
+		reaped += 1;
+		queue->base.outstanding -= 1;
+	}
+
+	if (queue_ret == QUEUE_ERROR) {
+		return -EIO;
+	}
+
+	return reaped;
+}
+
+int
+xnvme_be_macos_driverkit_async_cmd_io(struct xnvme_cmd_ctx *ctx, void *dbuf, size_t dbuf_nbytes,
+				      void *mbuf, size_t mbuf_nbytes)
+{
+	struct xnvme_queue_macos_driverkit *q =
+		(struct xnvme_queue_macos_driverkit *)ctx->async.queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)ctx->dev->be.state;
+	NvmeSubmitCmd cmd = {0};
+	int queue_ret;
+	struct buffer *buf;
+	uint64_t _buf_base_offset;
+
+	cmd.queue_id = state->qid_sync;
+
+	switch (ctx->cmd.common.opcode) {
+	case XNVME_SPEC_NVM_OPC_READ:
+		break;
```

Comment:
In my opinion there's no need for the empty cases in the switch statement

Reply:
Maybe, maybe not. It informs the next developer that it's a conscious decision to pass these through unchanged.

Reply:
Sure, but `ctx->cmd.common.opcode` has sufficiently many cases that it doesn't make sense to include the ones that don't require special handling.
That said, either way is fine.

Reply:
Then I'll prefer to keep it. To me, it serves a purpose.